zstash/globus.py

```diff
@@ -39,6 +43,23 @@
 archive_directory_listing: IterableTransferResponse = None
 
 
+def globus_flow(ep=""):
+    CLIENT = NativeAppAuthClient(
+        client_id="6c1629cf-446c-49e7-af95-323c6412397f", app_name="Zstash"
+    )
+    scopes = "urn:globus:auth:scope:transfer.api.globus.org:all"
+    endpoint_scope = f"[ *https://auth.globus.org/scopes/{ep}/data_access ]"
+    data_access_sopes = scopes + endpoint_scope
+    CLIENT.oauth2_start_flow(refresh_tokens=True, requested_scopes=data_access_sopes)
+    authorize_url = CLIENT.oauth2_get_authorize_url()
+    print(f"Please go to this URL and login:\n\n{authorize_url}\n\n")
+
+    get_input = getattr(__builtins__, "raw_input", input)
+    auth_code = get_input("Please enter the code you get after login here: ")
+    token_response = CLIENT.oauth2_exchange_code_for_tokens(auth_code)
+    print(token_response)
```

Comment:
@lukaszlacinski I'm trying to integrate @tylern4's `globus_flow` script (from https://github.com/E3SM-Project/zstash/issues/298#issuecomment-1728610607) into `zstash`. Then, users won't have to run it themselves before using `zstash` with Globus.

However, I'm getting quite confused as to where it belongs. I figured it should be part of the Globus activation, so I placed the calls in the `globus_activate` function of `zstash/globus.py`. 
- If I comment out the calls (and I have already granted consent manually), then `test_globus` works fine.
- If I include the calls, the test hangs on the `self.create(use_hpss, zstash_path, cache=self.cache)` line of the test (https://github.com/E3SM-Project/zstash/blob/main/tests/test_globus.py#L170)

I'm going to keep looking into this, but I figured I should ask if you had any immediate insights on how to automate the consent granting. Thanks!

Reply:
From what I can tell, the `auth_code = get_input("Please enter the code you get after login here: ")` line is causing the hang. 

However, many lines that should be printed before that point don't print, which made me think incorrectly that the error was occurring much earlier. From what I can tell, the `output, err = p.communicate()` line (https://github.com/E3SM-Project/zstash/blob/main/tests/base.py#L55) used to run `zstash create` just doesn't print any output until the entire command completes. And that of course includes the line asking for user input. 

I'm not quite sure how we would change that for testing purposes. In fact, we really don't want the unit tests to be asking for input anyway, as that would mean any test automation would also get stuck at that point.

I suppose a work-around would be to run a stand-alone `globus_flow` script before running the unit tests.

As for this pull request, I guess I can try testing the automated `globus_flow` functionality by manually transferring data to NERSC HPSS (from Perlmutter and from another machine).

Reply:
Ah ok, so it looks like we just need the `globus_flow(remote_endpoint)` line below, not the `globus_flow(local_endpoint)` line.

Manual test on Chrysalis:
```
# Create something to archive
$ emacs setup.sh 
mkdir zstash_demo
mkdir zstash_demo/empty_dir
mkdir zstash_demo/dir
echo 'file0 stuff' > zstash_demo/file0.txt
echo '' > zstash_demo/file_empty.txt
echo 'file1 stuff' > zstash_demo/dir/file1.txt

$ chmod 755 setup.sh
$ ./setup.sh
# Activate environment with `globus_flow` included in `zstash`
$ conda activate zstash_dev_n298
$ zstash create --hpss=globus://nersc/~/n298_chrysalis_dev zstash_demo

# On Perlmutter:
$ hsi
$ cd n298_chrysalis_dev
$ ls
# 000000.tar   index.db  
# Success

# Now, uncomment the `globus_flow(remote_endpoint)` line, and run `pip install .`
$ zstash create --hpss=globus://nersc/~/n298_chrysalis_dev_v2 zstash_demo
# This now asks for the consent (even though it should still be granted from an earlier run...). Once I enter the code, it completes.

# On Perlmutter:
$ hsi
$ cd n298_chrysalis_dev_v2
$ ls
# 000000.tar   index.db
# Success
```